recipes-sota/grub-ota-fallback/grub-ota-fallback/99_fallback_logic

```diff
@@ -0,0 +1,45 @@
+# GRUB can't add, so we hack an increment
+set inc_range="0,1 1,2 2,3"
+function inc_bootcount {
+    set incr=""
+    regexp --set=1:incr "${bootcount},([0-9]+)" "${inc_range}"
+    set bootcount=$incr
+}
+
+function inc_default {
+    set incr=""
+    regexp --set=1:incr "${default},([0-9]+)" "${inc_range}"
+    set default=$incr
+}
+
+# Load all the variables needed
+load_env
+
+if [ "${rollback}" == "1" -a "${upgrade_available}" = "1" ]; then
+    # Make sure to reset upgrade_available to avoid unnecessary wear
+    # Note this also makes rollback permanent. aktualizr will reset rollback
+    # when a new (hopefully better) update comes in.
+    set upgrade_available=0
+    save_env upgrade_available
+fi
+
+if [ "${bootcount}" == "${bootlimit}" ]; then
+    # We've reached our retry limit, marking for OS rollback
+    set rollback=1
+    save_env rollback
+else
+    inc_bootcount
+    save_env bootcount
+fi
+
+if [ "${upgrade_available}" == "1" ]; then
+    if [ "${rollback}" == 1 ]; then
+        # We need to rollback, so we reset 'bootcount' and increase default value.
+        # Since entries are ordered from newest to oldest, increasing default is
+        # rolling back to the last entry.
+        inc_default
```

Comment:
let's suppose that the user tries again, after a rollback, and then the update is successful. The `default` variable should be set to `0`?

Reply:
Like, user is on A, installs B via OTA, tries to boot onto B 3 times, fails all of them. Fallback kicks in, default changes to 1 and it boot onto A. Then user reboots and manually selects the B entry from GRUB menu?

Maybe this is an unsupported scenario, no? The user can definitely boot into the other image, but they would already be changing GRUB. I believe the default `timeout` is too fast for the user to select an entry. So if he already changed that, and for whatever reason was able to boot onto B and now thinks its a good idea to change the default boot to a previously failed image, he can also edit `grubenv` and change the value of default there.

What you had in mind?

Reply:
I'm thinking on, user is on A, install B via OTA, tries to boot onto B 3 times, fails all of them. Fallback kicks in, `default` changes to 1 and it boot onto A. Then, user is on A, try install C via OTA, it is successfull. But still the default continue being set as 1. Aktualizr does not seems to reset the `default` variable to 0.

Reply:
Right, after we talked I was able to understand this scenario.

Yes, in fact this was something that U-Boot would handle itself, and that I missed when doing our GRUB fallback script!
Now I've added a check at the top to reset `default` back to `0` when we have `rollback=0` and `update_available=1`. Since this state means that we have a new update deployed and ready to boot into, and since OSTree guarantees that every new deploy is always in priority `0`, we can safely do this.